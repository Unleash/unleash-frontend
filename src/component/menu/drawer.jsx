import React from "react";
import { Divider, Drawer, List, Icon } from "@material-ui/core";
import { NavLink } from "react-router-dom";
import classnames from "classnames";
import PropTypes from "prop-types";

import styles from "./drawer.module.scss";

import { baseRoutes as routes } from "./routes";

const filterByFlags = flags => r => {
    if (r.flag && !flags[r.flag]) {
        return false;
    }
    return true;
};

function getIcon(name) {
    if (name === "c_github") {
        return (
            <i
                className={classnames(
                    "material-icons",
                    styles.navigationIcon,
                    styles.iconGitHub
                )}
            />
        );
    } else {
        return <Icon className={styles.navigationIcon}>{name}</Icon>;
    }
}

function renderLink(link) {
    if (link.path) {
        return (
            <NavLink
                key={link.path}
                to={link.path}
                className={classnames(styles.navigationLink)}
                activeClassName={classnames(styles.navigationLinkActive)}
            >
                {getIcon(link.icon)} {link.value}
            </NavLink>
        );
    } else {
        return (
            <a
                href={link.href}
                key={link.href}
                target="_blank"
                className={[styles.navigationLink].join(" ")}
                title={link.title}
            >
                {getIcon(link.icon)} {link.value}
            </a>
        );
    }
}

export const DrawerMenu = ({
    links = [],
    title = "Unleash",
    flags = {},
    open = false,
    toggleDrawer
}) => (
    <Drawer
        className={styles.drawer}
        open={open}
        anchor={"left"}
        onClose={() => toggleDrawer()}
    >
        <div className={styles.drawerContainer}>
            <div className={styles.drawerTitleContainer}>
                <span className={[styles.drawerTitle].join(" ")}>
                    <img
                        src="public/logo.png"
                        width="32"
                        height="32"
                        className={styles.drawerTitleLogo}
                    />
                    <span className={styles.drawerTitleText}>{title}</span>
                </span>
            </div>
            <Divider />
            <List className={styles.drawerList}>
                {routes.filter(filterByFlags(flags)).map(item => (
                    <NavLink
                        onClick={() => toggleDrawer()}
                        key={item.path}
                        to={item.path}
                        className={classnames(styles.navigationLink)}
                        activeClassName={classnames(
                            styles.navigationLinkActive
                        )}
                    >
                        {getIcon(item.icon)}
                        {item.title}
                    </NavLink>
                ))}
            </List>
            <Divider />
            <List className={styles.navigation}>{links.map(renderLink)}</List>
        </div>
    </Drawer>
);

DrawerMenu.propTypes = {
    links: PropTypes.array,
    title: PropTypes.string,
    flags: PropTypes.object
};
