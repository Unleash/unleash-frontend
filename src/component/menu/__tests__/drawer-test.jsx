import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import { DrawerMenu } from '../drawer';

jest.mock('@material-ui/core');

test('should render DrawerMenu', () => {
    const tree = renderer.create(
        <MemoryRouter>
            <DrawerMenu />
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});

test('should render DrawerMenu with "features" selected', () => {
    const tree = renderer.create(
        <MemoryRouter initialEntries={['/features']}>
            <DrawerMenu />
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});
