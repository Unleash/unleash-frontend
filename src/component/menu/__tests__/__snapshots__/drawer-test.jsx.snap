// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DrawerMenu 1`] = `
<react-mdl-Drawer
  style={
    Object {
      "border": 0,
      "boxShadow": "none",
    }
  }
>
  <span
    className="drawerTitle mdl-layout-title"
  >
    <img
      className="drawerTitleLogo"
      height="32"
      src="public/logo.png"
      width="32"
    />
    <span
      className="drawerTitleText"
    >
      Unleash
    </span>
  </span>
  <hr />
  <react-mdl-Navigation
    className="navigation"
  >
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/features"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="list"
      />
       
      Feature Toggles
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/strategies"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="extension"
      />
       
      Strategies
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/history"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="history"
      />
       
      Event History
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/archive"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="archive"
      />
       
      Archived Toggles
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/applications"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="apps"
      />
       
      Applications
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/tag-types"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="label"
      />
       
      Tag types
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/logout"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="exit_to_app"
      />
       
      Sign out
    </a>
  </react-mdl-Navigation>
  <hr />
  <react-mdl-Navigation
    className="navigation"
  />
</react-mdl-Drawer>
`;

exports[`should render DrawerMenu with "features" selected 1`] = `
<react-mdl-Drawer
  style={
    Object {
      "border": 0,
      "boxShadow": "none",
    }
  }
>
  <span
    className="drawerTitle mdl-layout-title"
  >
    <img
      className="drawerTitleLogo"
      height="32"
      src="public/logo.png"
      width="32"
    />
    <span
      className="drawerTitleText"
    >
      Unleash
    </span>
  </span>
  <hr />
  <react-mdl-Navigation
    className="navigation"
  >
    <a
      aria-current="page"
      className="navigationLink mdl-color-text--grey-900 navigationLink mdl-color-text--black mdl-color--blue-grey-50"
      href="/features"
      onClick={[Function]}
      style={Object {}}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="list"
      />
       
      Feature Toggles
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/strategies"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="extension"
      />
       
      Strategies
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/history"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="history"
      />
       
      Event History
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/archive"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="archive"
      />
       
      Archived Toggles
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/applications"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="apps"
      />
       
      Applications
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/tag-types"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="label"
      />
       
      Tag types
    </a>
    <a
      aria-current={null}
      className="navigationLink mdl-color-text--grey-900"
      href="/logout"
      onClick={[Function]}
    >
      <react-mdl-Icon
        className="navigationIcon"
        name="exit_to_app"
      />
       
      Sign out
    </a>
  </react-mdl-Navigation>
  <hr />
  <react-mdl-Navigation
    className="navigation"
  />
</react-mdl-Drawer>
`;
